Lib9c/Extensions/EventRecipeExtensions.cs

```diff
@@ -44,5 +45,77 @@ public static class EventRecipeExtensions
                 .OrderBy(row => row.Id)
                 .ToList();
         }
+
+        public static EventMaterialItemRecipeSheet.Row ValidateFromAction(
+            this EventMaterialItemRecipeSheet sheet,
+            int eventMaterialItemRecipeId,
+            string actionTypeText,
+            string addressesHex)
+        {
+            if (!sheet.TryGetValue(eventMaterialItemRecipeId, out var row))
+            {
+                throw new InvalidActionFieldException(
+                    actionTypeText,
+                    addressesHex,
+                    nameof(eventMaterialItemRecipeId),
+                    " Aborted because the event recipe is not found.",
+                    new SheetRowNotFoundException(
+                        addressesHex,
+                        sheet.Name,
+                        eventMaterialItemRecipeId));
+            }
+
+            return row;
+        }
+
+        public static void ValidateFromAction(
+            this EventMaterialItemRecipeSheet.Row recipeRow,
+            MaterialItemSheet materialItemSheet,
+            Dictionary<int, int> materialsToUse,
+            string actionTypeText,
+            string addressesHex
+        )
+        {
+            var materialsCount = 0;
+            foreach (var pair in materialsToUse.OrderBy(pair => pair.Key))
+            {
+                if (!materialItemSheet.TryGetValue(pair.Key, out _))
+                {
+                    throw new SheetRowNotFoundException(
+                        addressesHex,
+                        nameof(MaterialItemSheet),
+                        pair.Key);
+                }
+
+                if (recipeRow.RequiredMaterialsId.Contains(pair.Key))
+                {
+                    materialsCount += pair.Value;
+                }
+            }
+
+            if (recipeRow.RequiredMaterialsCount != materialsCount)
+            {
+                throw new InvalidMaterialCountException(
+                    actionTypeText,
+                    addressesHex,
+                    recipeRow.RequiredMaterialsCount,
+                    materialsCount);
+            }
+        }
+
+        public static List<EventMaterialItemRecipeSheet.Row> GetRecipeRows(
+            this EventMaterialItemRecipeSheet sheet,
+            int eventScheduleId)
```

Comment:
where is this method used?

Reply:
It used in 9c client.

Reply:
If it's only used on the client, it would be better to move this method to the client side.